# Problem Statement
- This Dataset Contain the information about the bank customer that will take loan.
- Our goal is build a model that will detect weather the customer is eligible for loan or not.
- Data should be present in 2 different files.
- Both conatin same data about bank.
- Both files also contain `-99999` values.
- `-99999` are the missing values.

# Import Packages

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os
import warnings
warnings.filterwarnings('ignore')

# Read Data
1. Data is place another folder now find the path and read data.

In [2]:
file_path=[]
for i in os.listdir("Data"):
    file_path.append(f"{os.path.join(os.getcwd())}\Data\{i}")
    print(f"{os.path.join(os.getcwd())}\Data\{i}")

C:\Users\Laptop House\Weekly Task\CreditRisk\Data\case_study1.xlsx
C:\Users\Laptop House\Weekly Task\CreditRisk\Data\case_study2.xlsx


In [3]:
df1=pd.read_excel(file_path[0])
df2=pd.read_excel(file_path[1])

# Data Preprocessing
1. Top records
2. Shape of data
3. Check columns
4. Merge based on `common column`
5. Check Datatyes
6. Null values
7. Duplicates
8. Unique Values
9. Statistical Summary

# 1.Top Records

In [4]:
df1.sample(10)

PROSPECTID  Total_TL  Tot_Closed_TL  Tot_Active_TL  \
3016         3017         1              1              0   
8604         8605         3              2              1   
39877       39878         4              2              2   
27364       27365         1              0              1   
22351       22352         6              5              1   
9646         9647         1              0              1   
31662       31663         5              1              4   
31885       31886         2              0              2   
7607         7608         1              0              1   
6323         6324         6              3              3   

       Total_TL_opened_L6M  Tot_TL_closed_L6M  pct_tl_open_L6M  \
3016                     0                  0            0.000   
8604                     1                  0            0.333   
39877                    0                  2            0.000   
27364                    0                  0            0.000   
22351                    0                  0            0.000   
9646                     0                  0            0.000   
31662                    3                  1            0.600   
31885                    2                  0            1.000   
7607                     0                  0            0.000   
6323                     2                  0            0.333   

       pct_tl_closed_L6M  pct_active_tl  pct_closed_tl  ...  CC_TL  \
3016                 0.0          0.000          1.000  ...      0   
8604                 0.0          0.333          0.667  ...      0   
39877                0.5          0.500          0.500  ...      0   
27364                0.0          1.000          0.000  ...      0   
22351                0.0          0.167          0.833  ...      0   
9646                 0.0          1.000          0.000  ...      0   
31662                0.2          0.800          0.200  ...      0   
31885                0.0          1.000          0.000  ...      0   
7607                 0.0          1.000          0.000  ...      0   
6323                 0.0          0.500          0.500  ...      0   

       Consumer_TL  Gold_TL  Home_TL  PL_TL  Secured_TL  Unsecured_TL  \
3016             1        0        0      0           0             1   
8604             3        0        0      0           0             3   
39877            2        1        0      0           2             2   
27364            0        0        0      0           1             0   
22351            1        3        0      0           5             1   
9646             0        0        0      0           1             0   
31662            4        0        0      0           0             5   
31885            1        0        0      0           0             2   
7607             0        0        0      1           0             1   
6323             1        0        0      1           4             2   

       Other_TL  Age_Oldest_TL  Age_Newest_TL  
3016          0             41             41  
8604          0            106              4  
39877         0             14              7  
27364         0             30             30  
22351         1            116              6  
9646          0             10             10  
31662         1             14              3  
31885         1              5              5  
7607          0             26             26  
6323          0             69              5  

[10 rows x 26 columns]

In [5]:
df2.sample(10)

PROSPECTID  time_since_recent_payment  time_since_first_deliquency  \
27474       27475                       2431                       -99999   
45317       45318                        459                       -99999   
39410       39411                        103                       -99999   
15858       15859                         61                       -99999   
34039       34040                         42                       -99999   
963           964                         83                       -99999   
31609       31610                         36                           31   
48440       48441                       1923                       -99999   
12848       12849                         38                       -99999   
17653       17654                        823                       -99999   

       time_since_recent_deliquency  num_times_delinquent  \
27474                        -99999                     0   
45317                        -99999                     0   
39410                        -99999                     0   
15858                        -99999                     0   
34039                        -99999                     0   
963                          -99999                     0   
31609                             2                     9   
48440                        -99999                     0   
12848                        -99999                     0   
17653                        -99999                     0   

       max_delinquency_level  max_recent_level_of_deliq  num_deliq_6mts  \
27474                 -99999                          0               0   
45317                 -99999                          0               0   
39410                 -99999                          0               0   
15858                 -99999                          0               0   
34039                 -99999                          0               0   
963                   -99999                          0               0   
31609                     58                         30               4   
48440                 -99999                          0               0   
12848                 -99999                          0               0   
17653                 -99999                          0               0   

       num_deliq_12mts  num_deliq_6_12mts  ...  pct_CC_enq_L6m_of_L12m  \
27474                0                  0  ...                     0.0   
45317                0                  0  ...                     0.0   
39410                0                  0  ...                     0.0   
15858                0                  0  ...                     0.0   
34039                0                  0  ...                     0.0   
963                  0                  0  ...                     0.0   
31609                5                  1  ...                     0.0   
48440                0                  0  ...                     0.0   
12848                0                  0  ...                     0.0   
17653                0                  0  ...                     0.0   

       pct_PL_enq_L6m_of_ever  pct_CC_enq_L6m_of_ever  \
27474                     1.0                     0.0   
45317                     0.0                     0.0   
39410                     0.0                     0.0   
15858                     0.0                     0.0   
34039                     0.0                     0.0   
963                       0.0                     0.0   
31609                     0.0                     0.0   
48440                     0.0                     0.0   
12848                     0.0                     0.0   
17653                     0.0                     0.0   

       max_unsec_exposure_inPct  HL_Flag  GL_Flag  last_prod_enq2  \
27474                     0.103        1        0              PL   
45317                -99999.000        1        0    ConsumerLoan   
39410      

# Conclussion
1. we can see that in `df2` most of the columns contain `-99999` values.
2. These `-99999` are basically `Missing value`.

# 2.shape

In [6]:
print("Data 1 Shape: ",df1.shape)
print("Data 2 Shape: ",df2.shape)

Data 1 Shape:  (51336, 26)
Data 2 Shape:  (51336, 62)


# Conclussion 
1. Data1 contain `51336` rows and `26` Column
2. Data2 contain `51336` rows and `62` Column

# 3.Columns

In [7]:
print(df1.columns)
print(df2.columns)

Index(['PROSPECTID', 'Total_TL', 'Tot_Closed_TL', 'Tot_Active_TL',
       'Total_TL_opened_L6M', 'Tot_TL_closed_L6M', 'pct_tl_open_L6M',
       'pct_tl_closed_L6M', 'pct_active_tl', 'pct_closed_tl',
       'Total_TL_opened_L12M', 'Tot_TL_closed_L12M', 'pct_tl_open_L12M',
       'pct_tl_closed_L12M', 'Tot_Missed_Pmnt', 'Auto_TL', 'CC_TL',
       'Consumer_TL', 'Gold_TL', 'Home_TL', 'PL_TL', 'Secured_TL',
       'Unsecured_TL', 'Other_TL', 'Age_Oldest_TL', 'Age_Newest_TL'],
      dtype='object')
Index(['PROSPECTID', 'time_since_recent_payment',
       'time_since_first_deliquency', 'time_since_recent_deliquency',
       'num_times_delinquent', 'max_delinquency_level',
       'max_recent_level_of_deliq', 'num_deliq_6mts', 'num_deliq_12mts',
       'num_deliq_6_12mts', 'max_deliq_6mts', 'max_deliq_12mts',
       'num_times_30p_dpd', 'num_times_60p_dpd', 'num_std', 'num_std_6mts',
       'num_std_12mts', 'num_sub', 'num_sub_6mts', 'num_sub_12mts', 'num_dbt',
       'num_dbt_6mts', 'num_dbt_

# Conclussion
1. See that both `dataframe` one common column `PROSPECTID`.
2. We can `merge` them first and then perform other data preprocessing.

# 4.Merge `df1` anf `df2`

In [8]:
df1.head(2)

PROSPECTID  Total_TL  Tot_Closed_TL  Tot_Active_TL  Total_TL_opened_L6M  \
0           1         5              4              1                    0   
1           2         1              0              1                    0   

   Tot_TL_closed_L6M  pct_tl_open_L6M  pct_tl_closed_L6M  pct_active_tl  \
0                  0              0.0                0.0            0.2   
1                  0              0.0                0.0            1.0   

   pct_closed_tl  ...  CC_TL  Consumer_TL  Gold_TL  Home_TL  PL_TL  \
0            0.8  ...      0            0        1        0      4   
1            0.0  ...      0            1        0        0      0   

   Secured_TL  Unsecured_TL  Other_TL  Age_Oldest_TL  Age_Newest_TL  
0           1             4         0             72             18  
1           0             1         0              7              7  

[2 rows x 26 columns]

In [9]:
df2.head(2)

PROSPECTID  time_since_recent_payment  time_since_first_deliquency  \
0           1                        549                           35   
1           2                         47                       -99999   

   time_since_recent_deliquency  num_times_delinquent  max_delinquency_level  \
0                            15                    11                     29   
1                        -99999                     0                 -99999   

   max_recent_level_of_deliq  num_deliq_6mts  num_deliq_12mts  \
0                         29               0                0   
1                          0               0                0   

   num_deliq_6_12mts  ...  pct_CC_enq_L6m_of_L12m  pct_PL_enq_L6m_of_ever  \
0                  0  ...                     0.0                     0.0   
1                  0  ...                     0.0                     0.0   

   pct_CC_enq_L6m_of_ever  max_unsec_exposure_inPct  HL_Flag  GL_Flag  \
0                     0.0                    13.333        1        0   
1                     0.0                     0.860        0        0   

   last_prod_enq2  first_prod_enq2  Credit_Score  Approved_Flag  
0              PL               PL           696             P2  
1    ConsumerLoan     ConsumerLoan           685             P2  

[2 rows x 62 columns]

In [10]:
final_df=df1.merge(df2,how='inner',on='PROSPECTID')
final_df.head(2)

PROSPECTID  Total_TL  Tot_Closed_TL  Tot_Active_TL  Total_TL_opened_L6M  \
0           1         5              4              1                    0   
1           2         1              0              1                    0   

   Tot_TL_closed_L6M  pct_tl_open_L6M  pct_tl_closed_L6M  pct_active_tl  \
0                  0              0.0                0.0            0.2   
1                  0              0.0                0.0            1.0   

   pct_closed_tl  ...  pct_CC_enq_L6m_of_L12m  pct_PL_enq_L6m_of_ever  \
0            0.8  ...                     0.0                     0.0   
1            0.0  ...                     0.0                     0.0   

   pct_CC_enq_L6m_of_ever  max_unsec_exposure_inPct  HL_Flag  GL_Flag  \
0                     0.0                    13.333        1        0   
1                     0.0                     0.860        0        0   

   last_prod_enq2  first_prod_enq2  Credit_Score  Approved_Flag  
0              PL               PL           696             P2  
1    ConsumerLoan     ConsumerLoan           685             P2  

[2 rows x 87 columns]

# Conclussion
1. we have a `final df` now we can further work on this `final_df`

# 5. Datatypes

In [11]:
data_types=final_df.dtypes.to_dict()

In [12]:
pd.set_option('display.max_rows', None)  # Show all rows
pd.set_option('display.max_columns', None)  # Show all columns
pd.DataFrame(list(data_types.items()),columns=['Columns','Datatypes']).head(87)

Columns Datatypes
0                     PROSPECTID     int64
1                       Total_TL     int64
2                  Tot_Closed_TL     int64
3                  Tot_Active_TL     int64
4            Total_TL_opened_L6M     int64
5              Tot_TL_closed_L6M     int64
6                pct_tl_open_L6M   float64
7              pct_tl_closed_L6M   float64
8                  pct_active_tl   float64
9                  pct_closed_tl   float64
10          Total_TL_opened_L12M     int64
11            Tot_TL_closed_L12M     int64
12              pct_tl_open_L12M   float64
13            pct_tl_closed_L12M   float64
14               Tot_Missed_Pmnt     int64
15                       Auto_TL     int64
16                         CC_TL     int64
17                   Consumer_TL     int64
18                       Gold_TL     int64
19                       Home_TL     int64
20                         PL_TL     int64
21                    Secured_TL     int64
22                  Unsecured_TL     int64
23                      Other_TL     int64
24                 Age_Oldest_TL     int64
25                 Age_Newest_TL     int64
26     time_since_recent_payment     int64
27   time_since_first_deliquency     int64
28  time_since_recent_deliquency     int64
29          num_times_delinquent     int64
30         max_delinquency_level     int64
31     max_recent_level_of_deliq     int64
32                num_deliq_6mts     int64
33               num_deliq_12mts     int64
34             num_deliq_6_12mts     int64
35                max_deliq_6mts     int64
36               max_deliq_12mts     int64
37             num_times_30p_dpd     int64
38             num_times_60p_dpd     int64
39                       num_std     int64
40                  num_std_6mts     int64
41                 num_std_12mts     int64
42                       num_sub     int64
43                  num_sub_6mts     int64
44                 num_sub_12mts     int64
45                       num_dbt     int64
46                  num_dbt_6mts     int64
47                 num_dbt_12mts     int64
48                       num_lss     int64
49                  num_lss_6mts     int64
50                 num_lss_12mts     int64
51         recent_level_of_deliq     int64
52                       tot_enq     int64
53                        CC_enq     int64
54                    CC_enq_L6m     int64
55                   CC_enq_L12m     int64
56                        PL_enq     int64
57                    PL_enq_L6m     int64
58                   PL_enq_L12m     int64
59         time_since_recent_enq     int64
60                      enq_L12m     int64
61                       enq_L6m     int64
62                       enq_L3m     int64
63                 MARITALSTATUS    object
64                     EDUCATION    object
65                           AGE     int64
66                        GENDER    object
67              NETMONTHLYINCOME     int64
68           Time_With_Curr_Empr     int64
69        pct_of_active_TLs_ever   float64
70    pct_opened_TLs_L6m_of_L12m   float64
71         pct_currentBal_all_TL   float64
72                CC_utilization   float64
73                       CC_Flag     int64
74                PL_utilization   float64
75                       PL_Flag     int64
76        pct_PL_enq_L6m_of_L12m   float64
77        pct_CC_enq_L6m_of_L12m   float64
78        pct_PL_enq_L6m_of_ever   float64
79        pct_CC_enq_L6m_of_ever   float64
80      max_unsec_exposure_inPct   float64
81                       HL_Flag     int64
82                       GL_Flag     int64
83                last_prod_enq2    object
84               first_prod_enq2    object
85                  Credit_Score     int64
86                 Approved_Flag    object

In [13]:
# object colums need =['material_status','Education','gender','last_prod_loan','first_prod_loan','Approved_flag']
# int columns need =[All except object col]

In [14]:
final_df.select_dtypes('object').columns

Index(['MARITALSTATUS', 'EDUCATION', 'GENDER', 'last_prod_enq2',
       'first_prod_enq2', 'Approved_Flag'],
      dtype='object')

In [15]:
final_df.select_dtypes('number').columns

Index(['PROSPECTID', 'Total_TL', 'Tot_Closed_TL', 'Tot_Active_TL',
       'Total_TL_opened_L6M', 'Tot_TL_closed_L6M', 'pct_tl_open_L6M',
       'pct_tl_closed_L6M', 'pct_active_tl', 'pct_closed_tl',
       'Total_TL_opened_L12M', 'Tot_TL_closed_L12M', 'pct_tl_open_L12M',
       'pct_tl_closed_L12M', 'Tot_Missed_Pmnt', 'Auto_TL', 'CC_TL',
       'Consumer_TL', 'Gold_TL', 'Home_TL', 'PL_TL', 'Secured_TL',
       'Unsecured_TL', 'Other_TL', 'Age_Oldest_TL', 'Age_Newest_TL',
       'time_since_recent_payment', 'time_since_first_deliquency',
       'time_since_recent_deliquency', 'num_times_delinquent',
       'max_delinquency_level', 'max_recent_level_of_deliq', 'num_deliq_6mts',
       'num_deliq_12mts', 'num_deliq_6_12mts', 'max_deliq_6mts',
       'max_deliq_12mts', 'num_times_30p_dpd', 'num_times_60p_dpd', 'num_std',
       'num_std_6mts', 'num_std_12mts', 'num_sub', 'num_sub_6mts',
       'num_sub_12mts', 'num_dbt', 'num_dbt_6mts', 'num_dbt_12mts', 'num_lss',
       'num_lss_6mts', 'n

# Conclussion
1. `Datatypes` should be in correct format.

# 6. Null Values
1. First we can make the copy of dataframe

In [16]:
copy_df=final_df.copy()

In [17]:
final_df.isnull().sum()

PROSPECTID                      0
Total_TL                        0
Tot_Closed_TL                   0
Tot_Active_TL                   0
Total_TL_opened_L6M             0
Tot_TL_closed_L6M               0
pct_tl_open_L6M                 0
pct_tl_closed_L6M               0
pct_active_tl                   0
pct_closed_tl                   0
Total_TL_opened_L12M            0
Tot_TL_closed_L12M              0
pct_tl_open_L12M                0
pct_tl_closed_L12M              0
Tot_Missed_Pmnt                 0
Auto_TL                         0
CC_TL                           0
Consumer_TL                     0
Gold_TL                         0
Home_TL                         0
PL_TL                           0
Secured_TL                      0
Unsecured_TL                    0
Other_TL                        0
Age_Oldest_TL                   0
Age_Newest_TL                   0
time_since_recent_payment       0
time_since_first_deliquency     0
time_since_recent_deliquency    0
num_times_deli

1. **We can't see missing values but in above we say that `-99999` shows the missing values**
2. **Check if the ratio of `-99999` is greater in the specific column remove the columns**

# Handling Missing Values

In [18]:
final_df[final_df['time_since_recent_payment']==-99999].shape[0]

4291

In [19]:
final_df.shape

(51336, 87)

1. **We have `51336` records we can set the treshold `10000` if any columns contain `-99999` more then 10000 we can drop the columns.** 
2. **Also check the ratio of containg `-99999`.**

In [21]:
col_to_drop=[]
for i in final_df.columns:
    temp=final_df[final_df[i]==-99999].shape[0]
    ratio= round(temp/len(final_df),3)*100
    if(temp>10000):
        print(f"{i} -> -99999 value is -> {temp} and ratio is {ratio}%")
        col_to_drop.append(i)

time_since_first_deliquency -> -99999 value is -> 35949 and ratio is 70.0%
time_since_recent_deliquency -> -99999 value is -> 35949 and ratio is 70.0%
max_delinquency_level -> -99999 value is -> 35949 and ratio is 70.0%
max_deliq_6mts -> -99999 value is -> 12890 and ratio is 25.1%
max_deliq_12mts -> -99999 value is -> 10832 and ratio is 21.099999999999998%
CC_utilization -> -99999 value is -> 47636 and ratio is 92.80000000000001%
PL_utilization -> -99999 value is -> 44435 and ratio is 86.6%
max_unsec_exposure_inPct -> -99999 value is -> 23178 and ratio is 45.1%


# Conclussion
1. `time_since_first_deliquency` contain `70.0%` missing data.
2. `time_since_recent_deliquency` contain `70.0%` missing data.
3. `max_delinquency_level` contain `70.0%` missing data.
4. `max_deliq_6mts` contain `25.1%` missing data.
5. `max_deliq_12mts` contain  `21.09% missing data.
6. `CC_utilization` conrain `92.8%` missing data.
7. `PL_utilization` contain `86.6%` missing data.
8. `max_unsec_exposure_inPct` contain  `45.1%` missing data.
- **We will drop these columns**

### Drop Columns 

In [22]:
final_df.drop(columns=col_to_drop,inplace=True)

In [23]:
final_df.shape

(51336, 79)

In [24]:
final_df.sample(10)

PROSPECTID  Total_TL  Tot_Closed_TL  Tot_Active_TL  \
25555       25556         1              0              1   
25527       25528         8              1              7   
35245       35246         1              0              1   
43508       43509         4              1              3   
47045       47046         2              0              2   
26091       26092         1              0              1   
28339       28340         4              1              3   
49608       49609         2              1              1   
43815       43816         1              1              0   
47284       47285         1              0              1   

       Total_TL_opened_L6M  Tot_TL_closed_L6M  pct_tl_open_L6M  \
25555                    0                  0             0.00   
25527                    2                  1             0.25   
35245                    0                  0             0.00   
43508                    3                  1             0.75   
47045                    1                  0             0.50   
26091                    1                  0             1.00   
28339                    0                  0             0.00   
49608                    0                  0             0.00   
43815                    0                  0             0.00   
47284                    0                  0             0.00   

       pct_tl_closed_L6M  pct_active_tl  pct_closed_tl  Total_TL_opened_L12M  \
25555              0.000          1.000          0.000                     1   
25527              0.125          0.875          0.125                     5   
35245              0.000          1.000          0.000                     0   
43508              0.250          0.750          0.250                     3   
47045              0.000          1.000          0.000                     1   
26091              0.000          1.000          0.000                     1   
28339              0.000          0.750          0.250                     0   
49608              0.000          0.500          0.500                     1   
43815              0.000          0.000          1.000                     0   
47284              0.000          1.000          0.000                     0   

       Tot_TL_closed_L12M  pct_tl_open_L12M  pct_tl_closed_L12M  \
25555                   0             1.000               0.000   
25527                   1             0.625               0.125   
35245                   0             0.000               0.000   
43508                   1             0.750               0.250   
47045                   0             0.500               0.000   
26091                   0             1.000               0.000   
28339                   0             0.000               0.000   
49608                   1             0.500               0.500   
43815                   0             0.000               0.000   
47284                   0             0.000               0.000   

       Tot_Missed_Pmnt  Auto_TL  CC_TL  Consumer_TL  Gold_TL  Home_TL  PL_TL  \
25555                0        0      0            1        0        0      0   
25527                2        1      1            3        1        0      1   
35245                0        1      0            0        0        0      0   
43508                1        0      0            3        0        0      0   
47045                2        1      0            0        1        0      0   
26091                0        0      0            0        0        1      0   
28339                0        1      1            0        0        0      2   
49608                0        0      0            2        0        0      0   
43815                0        1      0            0        0        0      0   
47284                1        0      0            0        0        0      0   

       Secured_TL  Unsecured_TL  Other_TL  Age_Oldest_TL  Age_Newest_TL  \
25555           0             1  

# Conclussion
- After removing columns which contain `-99999` values more then `10000` shape of the dataframe is `51336` rows and `79` columns.
- we can remove only those columns which contain `-99999` values more then `10000` but remember that `-99999` are also present in those columns which contain  `-99999` values less then `10000`.
# Point to remember
- we have 2 option for `handling the remaing missing values`.
    ## Remove Missing Values
    - first we can remove those value.
        - we can remove the missing values if we maintain the balance of data we can remove them.
    ## Fill Missing Values
    - Fill these values
        - Filling the missing values is a cretical thing but if we can loss lot of data after removing the missing values then we pick the filling missing values.

- **Try to remove rows that contain `-99999` value**

In [25]:
for i in final_df.columns:
    print(final_df[~(final_df[i]==-99999)].shape[0])
    final_df=final_df[~(final_df[i]==-99999)]

51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51296
51296
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42064
42064
42064
42064
42064
42064
42064
42064
42064
42064
42064
42064
42064


# Conclussion
- Check the shape after removing.
- Also check the ration before removing and after removing

In [26]:
print(f"Shape of the data after removing -9999 values is: {final_df.shape}")

Shape of the data after removing -9999 values is: (42064, 79)


In [27]:
len(final_df)/len(copy_df)*100

81.93860059217702

# Conclussion
- we see that after removing the rows that contain `-99999` values we can maintain `80%` data.
- Now we sholud pick remove missing value and we will remove them.

In [28]:
# final_df.to_csv('credit.csv',index=False)

# Duplicates

In [29]:
final_df.duplicated().sum()

0

# Conclussion
- There is no duplicate found in this dataset.

# Unique Values

In [30]:
for i in final_df.describe(include='object'):
    print("Unique valeu in: ",i)
    
    print(final_df[i].unique())
    
    print("Total Unique values: ",final_df[i].nunique())
    
    print("-"*50)
    print()

Unique valeu in:  MARITALSTATUS
['Married' 'Single']
Total Unique values:  2
--------------------------------------------------

Unique valeu in:  EDUCATION
['12TH' 'GRADUATE' 'SSC' 'POST-GRADUATE' 'UNDER GRADUATE' 'OTHERS'
 'PROFESSIONAL']
Total Unique values:  7
--------------------------------------------------

Unique valeu in:  GENDER
['M' 'F']
Total Unique values:  2
--------------------------------------------------

Unique valeu in:  last_prod_enq2
['PL' 'ConsumerLoan' 'AL' 'CC' 'others' 'HL']
Total Unique values:  6
--------------------------------------------------

Unique valeu in:  first_prod_enq2
['PL' 'ConsumerLoan' 'others' 'AL' 'HL' 'CC']
Total Unique values:  6
--------------------------------------------------

Unique valeu in:  Approved_Flag
['P2' 'P1' 'P3' 'P4']
Total Unique values:  4
--------------------------------------------------



# Conclussion
- There are `2` unique values in  `MARITALSTATUS` i-e  `['Married' 'Single']`
- There are `7` unique values in  `EDUCATION` i-e `['12TH' 'GRADUATE' 'SSC' 'POST-GRADUATE' 'UNDER GRADUATE' 'OTHERS'
 'PROFESSIONAL']`
- There are `2` unique values in  `GENDER` i-e `['M' 'F']`
- There are `6` unique values in  `last_prod_enq2` i-e `['PL' 'ConsumerLoan' 'AL' 'CC' 'others' 'HL']`
- There are `6` unique values in  `first_prod_enq2` i-e `['PL' 'ConsumerLoan' 'others' 'AL' 'HL' 'CC']`
- There are `4` unique values in  `Approved_Flag` i-e `['P2' 'P1' 'P3' 'P4']`

# Statistical Summary

In [31]:
final_df.describe()

PROSPECTID      Total_TL  Tot_Closed_TL  Tot_Active_TL  \
count  42064.000000  42064.000000   42064.000000   42064.000000   
mean   25649.827477      5.262980       2.967383       2.295597   
std    14844.173396      7.463383       6.141098       2.404086   
min        1.000000      1.000000       0.000000       0.000000   
25%    12776.750000      1.000000       0.000000       1.000000   
50%    25706.500000      3.000000       1.000000       2.000000   
75%    38518.250000      6.000000       3.000000       3.000000   
max    51336.000000    235.000000     216.000000      47.000000   

       Total_TL_opened_L6M  Tot_TL_closed_L6M  pct_tl_open_L6M  \
count         42064.000000        42064.00000     42064.000000   
mean              0.812643            0.48992         0.179032   
std               1.383559            1.05892         0.278043   
min               0.000000            0.00000         0.000000   
25%               0.000000            0.00000         0.000000   
50%               0.000000            0.00000         0.000000   
75%               1.000000            1.00000         0.333000   
max              27.000000           19.00000         1.000000   

       pct_tl_closed_L6M  pct_active_tl  pct_closed_tl  Total_TL_opened_L12M  \
count       42064.000000   42064.000000   42064.000000          42064.000000   
mean            0.097783       0.577452       0.422548              1.672142   
std             0.210957       0.366110       0.366110              2.249543   
min             0.000000       0.000000       0.000000              0.000000   
25%             0.000000       0.286000       0.000000              0.000000   
50%             0.000000       0.545000       0.455000              1.000000   
75%             0.100000       1.000000       0.714000              2.000000   
max             1.000000       1.000000       1.000000             34.000000   

       Tot_TL_closed_L12M  pct_tl_open_L12M  pct_tl_closed_L12M  \
count        42064.000000      42064.000000        42064.000000   
mean             0.825504          0.401271            0.160365   
std              1.537208          0.381266            0.258831   
min              0.000000          0.000000            0.000000   
25%              0.000000          0.000000            0.000000   
50%              0.000000          0.333000            0.000000   
75%              1.000000          0.714000            0.250000   
max             33.000000          1.000000            1.000000   

       Tot_Missed_Pmnt       Auto_TL         CC_TL   Consumer_TL  \
count     42064.000000  42064.000000  42064.000000  42064.000000   
mean          0.525746      0.667626      0.145921      1.362400   
std           1.106442      0.952677      0.549314      2.394966   
min           0.000000      0.000000      0.000000      0.000000   
25%           0.000000      0.000000      0.000000      0.000000   
50%           0.000000      0.000000      0.000000      1.000000   
75%           1.000000      1.000000      0.000000      2.000000   
max          34.000000     27.000000     27.000000     41.000000   

            Gold_TL       Home_TL         PL_TL    Secured_TL  Unsecured_TL  \
count  42064.000000  42064.000000  42064.000000  42064.000000  42064.000000   
mean       1.566304      0.076241      0.328000      2.921334      2.341646   
std        5.500184      0.358582      0.916368      6.379764      3.405397   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      1.000000      1.000000   
75%        1.000000      0.000000      0.000000      3.000000      3.000000   
max      235.000000     10.000000     29.000000    235.000000     55.000000   

           Other_TL  Age_Oldest_TL  Age_Newest_TL  time_since_recent_payment  \
count  42064.000000   42064.000000   42064.000000               42064.000000   


# Conclussion
- **We can't see clear information first we can do `Feature Engnering` and then we can see clear information.**

# Feature Engnering